### PR TITLE
JDK-8281460: Let ObjectMonitor have its own NMT category

### PR DESCRIPTION
**Pull request by tstuefe**
It would be useful to distinguish ObjectMonitor allocations in NMT, as a simple way to both to see the needed footprint as well as the number of allocated monitors.

Note, this needs [JDK-8281450](https://bugs.openjdk.java.net/browse/JDK-8281450) as a prerequisite.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281460](https://bugs.openjdk.java.net/browse/JDK-8281460): Let ObjectMonitor have its own NMT category

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 121811e9c5966044246498f02270d251a57aab15
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to 121811e9c5966044246498f02270d251a57aab15
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 121811e9c5966044246498f02270d251a57aab15

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7378/head:pull/7378` \
`$ git checkout pull/7378`

Update a local copy of the PR: \
`$ git checkout pull/7378` \
`$ git pull https://git.openjdk.java.net/jdk pull/7378/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7378`

View PR using the GUI difftool: \
`$ git pr show -t 7378`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7378.diff">https://git.openjdk.java.net/jdk/pull/7378.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Hi Thomas,

Looks fine as far as it goes, but leads me to wonder about related data structures e.g. `ObjectMonitorsHashtable`, which seems to use `mtThread`. ??

Thanks,
David

**Comment by tstuefe** [comment]
> Hi Thomas,
> 
> Looks fine as far as it goes, but leads me to wonder about related data structures e.g. `ObjectMonitorsHashtable`, which seems to use `mtThread`. ??
> 
> Thanks, David

Hi David,

my intent was specifically to get a look at the number of outstanding ObjectMonitor instances. Do you think I am misusing NMT here? If yes, I'm fine with withdrawing the PR.

Cheers, Thomas

**Comment by tstuefe** [comment]
> Introducing a category just to collect info on a specific kind of allocation does seem somewhat fragile. I can easily imagine someone looking at ObjectMonitorHashtable and thinking it should be tagged too. But in the meantime if this serves a purpose ...

It feels weird now :) but let us see if someone else objects. If yes, I withdraw this change.

**Comment by shipilev** [comment]
> It feels weird now :) but let us see if someone else objects. If yes, I withdraw this change.

Yeah, that looks a bit weird, and I have another question: why aren't both `ObjectMonitor` and `ObjectMonitorsHashtable` are `mtSynchronizer`?

**Comment by tstuefe** [comment]
> > It feels weird now :) but let us see if someone else objects. If yes, I withdraw this change.
> 
> Yeah, that looks a bit weird, and I have another question: why aren't both `ObjectMonitor` and `ObjectMonitorsHashtable` are `mtSynchronizer`?

No idea. My guess is that ObjectMonitor is mtInternal because of an oversight in JDK-8253064 and the hashtable is only used as an aid during thread dumps, so mtThread? Would be a question to @dcubed-ojdk.

**Comment by dholmes-ora** [comment]
I think mtSynchronizer was for VM synchronization objects: mutexes, semaphore etc. Not ObjectMonitors. But as always if these conventions are not clearly known then "misuse" can creep in.

**Comment by dcubed-ojdk** [comment]
Actually `ObjectMonitor` is `mtInternal` in JDK-8253064 intentionally. Here's the original
code that allocated blocks of ObjectMonitors:

```
    void* real_malloc_addr = NEW_C_HEAP_ARRAY(char, aligned_size, mtInternal);
    temp = (PaddedObjectMonitor*)align_up(real_malloc_addr, OM_CACHE_LINE_SIZE);
    (void)memset((void *) temp, 0, neededsize);
```

so the change in allocation implementation in JDK-8253064 is consistent (NMT tag wise)
with what was there before.

That said, I have no complaints about using a specific NMT tag. However, I would
have used "ObjectMonitors" instead of "Monitors" to avoid confusion with VM internal
Monitors. One good reason for this change is that NMT is used for analyzing unexpected
memory usages and since we changed the allocation implementation and some of the
policy in JDK-8253064 it would be good to "highlight" that in NMT as we continue to
shake out what we did in JDK-8253064.

File: src/hotspot/share/memory/allocation.hpp [hunk 1]
```diff
@@ -146,6 +146,7 @@ class AllocatedObj {
   f(mtServiceability, "Serviceability")                                              \
   f(mtMetaspace,      "Metaspace")                                                   \
   f(mtStringDedup,    "String Deduplication")                                        \
+  f(mtObjectMonitor,  "Monitors")                                                    \
```
Comment (dcubed-ojdk) [inline src/hotspot/share/memory/allocation.hpp:149]: I would use "ObjectMonitors" in the string to distinguish from VM internal Monitors.
It also matches the `mtObjectMonitor` name better.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. I have one minor string rename request that
you can choose to ignore.

**Comment by tstuefe** [comment]
> Thumbs up. I have one minor string rename request that you can choose to ignore.

Thanks Dan! Your proposal makes sense, I'll change the string before pushing.

**Comment by tstuefe** [comment]
@dholmes-ora @shipilev Dan confirmed my original idea makes sense, are you okay with this change too?

..Thomas

**Review by shipilev — APPROVED** [review]
All right, but I also keep thinking that NMT becomes a wide-spread diagnostic interface, and at some point in the future we should probably exercise a bit of restraint in introducing overly-fine-grained categories.

**Review by dholmes-ora — APPROVED** [review]
I'm fine with this as is (modulo Dan's suggested ObjectMonitor change).

I agree with @shipilev that as NMT usage expands we are going to need to more carefully consider how we tag things (and whether we need a way to group tags) so that we can try to address fine- and coarse-grained enquiries.

Cheers,
David

**Comment by tstuefe** [comment]
Alexey:
> All right, but I also keep thinking that NMT becomes a wide-spread diagnostic interface, and at some point in the future we should probably exercise a bit of restraint in introducing overly-fine-grained categories.

David:
> I'm fine with this as is (modulo Dan's suggested ObjectMonitor change).
> 
> I agree with @shipilev that as NMT usage expands we are going to need to more carefully consider how we tag things (and whether we need a way to group tags) so that we can try to address fine- and coarse-grained inquiries.

I agree with both of you. I feel the situation is a bit like with UL tags. We introduced them, but at some point stopped thinking about rules. At least I am not aware of any rules. The result is a bit anarchic. With NMT it is less severe though since with UL, due to the fact how tags are combined when forming log expressions, once they exist you never can change them. NMT categories have fewer backward compatibility headaches.

The obvious exception are people who compare NMT output between JVM releases and get confused about numbers that changed. But that's just education.

NMT group tags were mentioned several times in other PRs, and I think that makes sense and is not that hard to do. Especially when done at a shallow level, just as a display thing (e.g. log with fine granularity categories a,b,c,d, then define "mtInternal" as a group of a,b,c,d, which only has to interest the NMT printing code).

**Pull request closed without merging**